ecc/bls12-381/sswu-fp.go

```diff
@@ -0,0 +1,314 @@
+func sqrtRatio(z *fp.Element, u *fp.Element, v *fp.Element) bool {
+	var tv1 fp.Element
+	tv1.Square(v)
+	var tv2 fp.Element
+	tv2.Mul(u, v)
+	tv1.Mul(&tv1, &tv2)
+
+	var y1 fp.Element
+	{
+		var c1 big.Int
+		// c1 = 1000602388805416848354447456433976039139220704984751971333014534031007912622709466110671907282253916009473568139946
+		c1.SetBytes([]byte{6, 128, 68, 122, 142, 95, 249, 166, 146, 198, 233, 237, 144, 210, 235, 53, 217, 29, 210, 225, 60, 225, 68, 175, 217, 204, 52, 168, 61, 172, 61, 137, 7, 170, 255, 255, 172, 84, 255, 255, 238, 127, 191, 255, 255, 255, 234, 170})
+		y1.Exp(tv1, &c1)
+	}
+
+	y1.Mul(&y1, &tv2)
+
+	var y2 fp.Element
+	y2.Mul(&y1, &fp.Element{17544630987809824292, 17306709551153317753, 8299808889594647786, 5930295261504720397, 675038575008112577, 167386374569371918})
+
+	var tv3 fp.Element
+	tv3.Square(&y1)
+	tv3.Mul(&tv3, v)
+	isQr := tv3 == *u
+
+	if isQr {
+		*z = y1
+	} else {
+		*z = y2
+	}
+
+	return isQr
+}
+
+//TODO: Use addchain
+//TODO: Might duplicate functionality from mulByConst functions
+// mulByZ multiplies x by 11 and stores the result in z
+func mulByZ(z *fp.Element, x *fp.Element) {
+
+	res := *x
+
+	res.Double(&res)
+
+	res.Double(&res)
+
+	res.Add(&res, x)
+
+	res.Double(&res)
+
+	res.Add(&res, x)
+
+	*z = res
+}
+
+// From https://datatracker.ietf.org/doc/draft-irtf-cfrg-hash-to-curve/13/ Pg 80
+func sswuMapG1(u *fp.Element) G1Affine {
+
+	var tv1 fp.Element
+	tv1.Square(u)
+
+	//mul tv1 by Z
+	mulByZ(&tv1, &tv1)
+
+	var tv2 fp.Element
+	tv2.Square(&tv1)
+	tv2.Add(&tv2, &tv1)
+
+	var tv3 fp.Element
+	//Standard doc line 5
+	var tv4 fp.Element
+	tv4.SetOne()
+	tv3.Add(&tv2, &tv4)
+	tv3.Mul(&tv3, &fp.Element{18129637713272545760, 11144507692959411567, 10108153527111632324, 9745270364868568433, 14587922135379007624, 469008097655535723})
+
+	tv4 = fp.Element{3415322872136444497, 9675504606121301699, 13284745414851768802, 2873609449387478652, 2897906769629812789, 1536947672689614213}
+	//TODO: Std doc uses conditional move. If-then-else good enough here?
+	if tv2.IsZero() {
+		mulByZ(&tv4, &tv4)
+		//WARNING: this branch takes less time
+	} else {
+		tv4.Mul(&tv4, &tv2)
+		tv4.Neg(&tv4)
+	}
+	tv2.Square(&tv3)
+
+	var tv6 fp.Element
+	//Standard doc line 10
+	tv6.Square(&tv4)
+
+	var tv5 fp.Element
+	tv5.Mul(&tv6, &fp.Element{3415322872136444497, 9675504606121301699, 13284745414851768802, 2873609449387478652, 2897906769629812789, 1536947672689614213})
+
+	tv2.Add(&tv2, &tv5)
+	tv2.Mul(&tv2, &tv3)
+	tv6.Mul(&tv6, &tv4)
+
+	//Standards doc line 15
+	tv5.Mul(&tv6, &fp.Element{18129637713272545760, 11144507692959411567, 10108153527111632324, 9745270364868568433, 14587922135379007624, 469008097655535723})
+	tv2.Add(&tv2, &tv5)
+
+	var x fp.Element
+	x.Mul(&tv1, &tv3)
+
+	var y1 fp.Element
+	gx1Square := sqrtRatio(&y1, &tv2, &tv6)
+
+	var y fp.Element
+	y.Mul(&tv1, u)
+
+	//Standards doc line 20
+	y.Mul(&y, &y1)
+
+	//TODO: Not constant time. Is it okay?
```

Comment:
ideally we would want a constant-time algorithm (but many algorithms in gnark-crypto are not)

Reply:
Conditional move is in the works!